File: assets/lua/meta/sound.lua
--- @meta
--- 
--- @class sound
sound = {}


--- 
--- @class sound.Source
sound.Source = {}

--- 
--- @param name string
--- @param gain number
--- @return bool ret0
--- @overload fun(name: string, gain: number, gainHF: number, gainLF: number): bool
function sound.Source:SetEffectParameters(name, gain) end

--- 
--- @overload fun(arg1: ents.BaseEntityBase): 
function sound.Source:SetSource() end

--- 
--- @return ents.BaseEntityBase ret0
function sound.Source:GetSource() end

--- 
--- @return number ret0
function sound.Source:GetOuterConeGainHF() end

--- 
--- @return number ret0
function sound.Source:GetGain() end

--- 
--- @param arg1 number
function sound.Source:SetOuterConeAngle(arg1) end

--- 
--- @param arg1 number
function sound.Source:SetOffset(arg1) end

--- 
--- @return math.Vector ret0
function sound.Source:GetVelocity() end

--- 
--- @param arg1 math.Vector
function sound.Source:SetVelocity(arg1) end

--- 
--- @return bool ret0
function sound.Source:IsLooping() end

--- 
--- @return bool ret0
function sound.Source:GetSendGainHFAuto() end

--- 
--- @param arg1 string
--- @param arg2 number
function sound.Source:SetEffectGain(arg1, arg2) end

--- 
--- @return bool ret0
function sound.Source:IsStopped() end

--- 
--- @return number ret0
function sound.Source:GetRolloffFactor() end

--- 
--- @return enum ALState ret0
function sound.Source:GetState() end

--- 
--- @return number ret0
function sound.Source:GetMaxGain() end

--- 
--- @param arg1 number
--- @param arg2 number
function sound.Source:SetRange(arg1, arg2) end

--- 
function sound.Source:Rewind() end

--- 
--- @param arg1 math.Vector
function sound.Source:SetDirection(arg1) end

--- 
--- @return number ret0_1
--- @return number ret0_2
function sound.Source:GetRangeOffsets() end

--- 
--- @return number ret0
function sound.Source:GetOffset() end

--- 
--- @param arg1 bool
function sound.Source:SetRelative(arg1) end

--- 
--- @return number ret0
function sound.Source:GetRoomRolloffFactor() end

--- 
--- @param arg1 int
function sound.Source:SetPriority(arg1) end

--- 
--- @param arg1 bool
--- @param arg2 bool
--- @param arg3 bool
function sound.Source:SetGainAuto(arg1, arg2, arg3) end

--- 
--- @return bool ret0
function sound.Source:IsRelative() end

--- 
--- @return number ret0
function sound.Source:GetReferenceDistance() end

--- 
--- @param arg1 number
function sound.Source:SetRoomRolloffFactor(arg1) end

--- 
--- @return int ret0
function sound.Source:GetPriority() end

--- 
--- @param arg1 number
function sound.Source:SetPitch(arg1) end

--- 
--- @return number ret0
function sound.Source:GetMaxDistance() end

--- 
--- @return bool ret0
function sound.Source:HasRange() end

--- 
--- @return number ret0
function sound.Source:GetMinGain() end

--- 
--- @param arg1 number
function sound.Source:SetRolloffFactor(arg1) end

--- 
--- @return number ret0
function sound.Source:GetOuterConeGain() end

--- 
--- @param arg1 number
function sound.Source:SetOuterConeGain(arg1) end

--- 
--- @param arg1 number
function sound.Source:SetOuterConeGainHF(arg1) end

--- 
--- @param arg1 number
--- @param arg2 number
function sound.Source:SetConeAngles(arg1, arg2) end

--- 
--- @param arg1 number
function sound.Source:SetMaxDistance(arg1) end

--- 
--- @param arg1 number
function sound.Source:SetMaxGain(arg1) end

--- 
--- @param arg1 enum ALSoundType
function sound.Source:AddType(arg1) end

--- 
--- @param arg1 number
function sound.Source:SetGain(arg1) end

--- 
--- @param arg2 unknown
--- @return util.Callback ret0
function sound.Source:CallOnStateChanged(arg2) end

--- 
--- @return bool ret0_1
--- @return bool ret0_2
--- @return bool ret0_3
function sound.Source:GetGainAuto() end

--- 
--- @param arg1 number
function sound.Source:FadeIn(arg1) end

--- 
--- @param arg1 number
function sound.Source:FadeOut(arg1) end

--- 
--- @return bool ret0
function sound.Source:GetSendGainAuto() end

--- 
--- @return number ret0_1
--- @return number ret0_2
function sound.Source:GetGainRange() end

--- 
--- @return number ret0_1
--- @return number ret0_2
function sound.Source:GetDistanceRange() end

--- 
--- @return number ret0
function sound.Source:GetMaxAudibleDistance() end

--- 
--- @param arg1 number
function sound.Source:SetFadeInDuration(arg1) end

--- 
--- @return number ret0_1
--- @return number ret0_2
function sound.Source:GetRange() end

--- 
--- @param gain number
--- @overload fun(gain: number, gainHF: number, gainLF: number): 
function sound.Source:SetDirectFilter(gain) end

--- 
--- @param arg1 string
function sound.Source:RemoveEffect(arg1) end

--- 
--- @return number ret0
function sound.Source:GetDuration() end

--- 
--- @return number ret0
function sound.Source:GetOuterConeAngle() end

--- 
--- @param arg1 number
function sound.Source:SetInnerConeAngle(arg1) end

--- 
--- @param arg1 number
--- @param arg2 number
function sound.Source:SetGainRange(arg1, arg2) end

--- 
--- @return number ret0
function sound.Source:GetPitch() end

--- 
--- @return number ret0
function sound.Source:GetLeftStereoAngle() end

--- 
--- @param arg1 number
function sound.Source:SetRightStereoAngle(arg1) end

--- 
--- @return number ret0
function sound.Source:GetRightStereoAngle() end

--- 
--- @param arg1 number
function sound.Source:SetAirAbsorptionFactor(arg1) end

--- 
--- @return number ret0
function sound.Source:GetAirAbsorptionFactor() end

--- 
--- @return math.Vector ret0
function sound.Source:GetDirection() end

--- 
--- @return number ret0_1
--- @return number ret0_2
function sound.Source:GetConeAngles() end

--- 
--- @param name string
--- @param gain number
--- @return bool ret0
--- @overload fun(name: string, gain: number, gainHF: number, gainLF: number): bool
function sound.Source:AddEffect(name, gain) end

--- 
--- @param arg1 number
--- @param arg2 number
function sound.Source:SetDistanceRange(arg1, arg2) end

--- 
--- @param arg1 number
function sound.Source:SetMinGain(arg1) end

--- 
--- @param arg1 number
function sound.Source:SetTimeOffset(arg1) end

--- 
--- @return number ret0
function sound.Source:GetTimeOffset() end

--- 
--- @param arg1 number
function sound.Source:SetReferenceDistance(arg1) end

--- 
--- @return number ret0_1
--- @return number ret0_2
function sound.Source:GetRolloffFactors() end

--- 
--- @param arg1 number
--- @param arg2 number
function sound.Source:SetRolloffFactors(arg1, arg2) end

--- 
--- @param arg1 bool
function sound.Source:SetLooping(arg1) end

--- 
--- @param arg1 number
--- @param arg2 number
function sound.Source:SetOuterConeGains(arg1, arg2) end

--- 
function sound.Source:Play() end

--- 
function sound.Source:Stop() end

--- 
function sound.Source:Pause() end

--- 
--- @return bool ret0
function sound.Source:GetDirectGainHFAuto() end

--- 
--- @return bool ret0
function sound.Source:IsPlaying() end

--- 
--- @return bool ret0
function sound.Source:IsPaused() end

--- 
--- @return number ret0_1
--- @return number ret0_2
function sound.Source:GetStereoAngles() end

--- 
--- @param arg1 number
--- @param arg2 number
function sound.Source:SetStereoAngles(arg1, arg2) end

--- 
--- @return number ret0_1
--- @return number ret0_2
function sound.Source:GetOuterConeGains() end

--- 
--- @return number ret0_1
--- @return number ret0_2
--- @return number ret0_3
function sound.Source:GetDirectFilter() end

--- 
--- @param arg1 number
function sound.Source:SetLeftStereoAngle(arg1) end

--- 
--- @return number ret0
function sound.Source:GetDopplerFactor() end

--- 
--- @param arg1 number
function sound.Source:SetDopplerFactor(arg1) end

--- 
--- @return math.Vector ret0_1
--- @return math.Vector ret0_2
function sound.Source:GetOrientation() end

--- 
function sound.Source:__tostring() end

--- 
--- @param arg1 math.Vector
--- @param arg2 math.Vector
function sound.Source:SetOrientation(arg1, arg2) end

--- 
--- @return number ret0
function sound.Source:GetFadeOutDuration() end

--- 
--- @return number ret0
function sound.Source:GetFadeInDuration() end

--- 
--- @param arg1 number
function sound.Source:SetFadeOutDuration(arg1) end

--- 
--- @param arg1 enum ALSoundType
function sound.Source:SetType(arg1) end

--- 
--- @return enum ALSoundType ret0
function sound.Source:GetType() end

--- 
--- @param arg1 math.Vector
function sound.Source:SetPos(arg1) end

--- 
--- @return math.Vector ret0
function sound.Source:GetPos() end

--- 
--- @return number ret0
function sound.Source:GetInnerConeAngle() end

--- 
--- @param arg1 math.Vector
--- @return number ret0
function sound.Source:GetIntensity(arg1) end

--- 
function sound.Source:ClearRange() end

--- 
--- @return table ret0
function sound.Source:GetPhonemeData() end


--- 
--- @class sound.Effect
sound.Effect = {}

--- 
--- @param props sound.Effect.EqualizerProperties
--- @overload fun(props: sound.Effect.CompressorProperties): 
--- @overload fun(props: sound.Effect.AutoWahProperties): 
--- @overload fun(props: sound.Effect.RingModulatorProperties): 
--- @overload fun(props: sound.Effect.PitchShifterProperties): 
--- @overload fun(props: sound.Effect.VocalMorpherProperties): 
--- @overload fun(props: sound.Effect.FrequencyShifterProperties): 
--- @overload fun(props: sound.Effect.FlangerProperties): 
--- @overload fun(props: sound.Effect.EchoProperties): 
--- @overload fun(props: sound.Effect.DistortionProperties): 
--- @overload fun(props: sound.Effect.ChorusProperties): 
--- @overload fun(props: sound.Effect.ReverbProperties): 
function sound.Effect:SetProperties(props) end


--- 
--- @class sound.Effect.PitchShifterProperties
--- @field coarseTune int 
--- @field fineTune int 
--- @overload fun():sound.Effect.PitchShifterProperties
sound.Effect.PitchShifterProperties = {}


--- 
--- @class sound.Effect.CompressorProperties
--- @field onOff int 
--- @overload fun():sound.Effect.CompressorProperties
sound.Effect.CompressorProperties = {}


--- 
--- @class sound.Effect.DistortionProperties
--- @field edge number 
--- @field lowpassCutoff number 
--- @field eqCenter number 
--- @field gain number 
--- @field eqBandwidth number 
--- @overload fun():sound.Effect.DistortionProperties
sound.Effect.DistortionProperties = {}


--- 
--- @class sound.Effect.AutoWahProperties
--- @field attackTime number 
--- @field releaseTime number 
--- @field resonance number 
--- @field peakGain number 
--- @overload fun():sound.Effect.AutoWahProperties
sound.Effect.AutoWahProperties = {}


--- 
--- @class sound.Effect.FlangerProperties
--- @field feedback number 
--- @field waveform int 
--- @field depth number 
--- @field delay number 
--- @field phase int 
--- @field rate number 
--- @overload fun():sound.Effect.FlangerProperties
sound.Effect.FlangerProperties = {}


--- 
--- @class sound.Effect.EqualizerProperties
--- @field mid2Width number 
--- @field mid1Gain number 
--- @field highGain number 
--- @field highCutoff number 
--- @field mid1Width number 
--- @field lowCutoff number 
--- @field lowGain number 
--- @field mid1Center number 
--- @field mid2Gain number 
--- @field mid2Center number 
--- @overload fun():sound.Effect.EqualizerProperties
sound.Effect.EqualizerProperties = {}


--- 
--- @class sound.Effect.VocalMorpherProperties
--- @field phonemeA int 
--- @field phonemeACoarseTuning int 
--- @field phonemeB int 
--- @field phonemeBCoarseTuning int 
--- @field waveform int 
--- @field rate number 
--- @overload fun():sound.Effect.VocalMorpherProperties
sound.Effect.VocalMorpherProperties = {}


--- 
--- @class sound.Effect.ChorusProperties
--- @field feedback number 
--- @field waveform int 
--- @field depth number 
--- @field delay number 
--- @field phase int 
--- @field rate number 
--- @overload fun():sound.Effect.ChorusProperties
sound.Effect.ChorusProperties = {}


--- 
--- @class sound.Effect.ReverbProperties
--- @field lfReference number 
--- @field modulationDepth number 
--- @field diffusion number 
--- @field gainHF number 
--- @field roomRolloffFactor number 
--- @field decayHFLimit int 
--- @field gain number 
--- @field density number 
--- @field gainLF number 
--- @field decayTime number 
--- @field decayHFRatio number 
--- @field decayLFRatio number 
--- @field reflectionsGain number 
--- @field reflectionsDelay number 
--- @field reflectionsPan math.Vector 
--- @field lateReverbDelay number 
--- @field lateReverbGain number 
--- @field lateReverbPan math.Vector 
--- @field echoTime number 
--- @field echoDepth number 
--- @field modulationTime number 
--- @field airAbsorptionGainHF number 
--- @field hfReference number 
--- @overload fun():sound.Effect.ReverbProperties
sound.Effect.ReverbProperties = {}


--- 
--- @class sound.Effect.FrequencyShifterProperties
--- @field frequency number 
--- @field leftDirection int 
--- @field rightDirection int 
--- @overload fun():sound.Effect.FrequencyShifterProperties
sound.Effect.FrequencyShifterProperties = {}


--- 
--- @class sound.Effect.EchoProperties
--- @field delay number 
--- @field spread number 
--- @field feedback number 
--- @field lrDelay number 
--- @field damping number 
--- @overload fun():sound.Effect.EchoProperties
sound.Effect.EchoProperties = {}


--- 
--- @class sound.Effect.RingModulatorProperties
--- @field frequency number 
--- @field waveform int 
--- @field highpassCutoff number 
--- @overload fun():sound.Effect.RingModulatorProperties
sound.Effect.RingModulatorProperties = {}

File: core/shared/src/implementation/lua/classes/sound.cpp
// SPDX-FileCopyrightText: (c) 2019 Silverlan <opensource@pragma-engine.com>
// SPDX-License-Identifier: MIT
module;

module pragma.shared;

import :scripting.lua.classes.sound;

namespace Lua {
	namespace ALSound {
		static CallbackHandle CallOnStateChanged(lua::State *l, pragma::audio::ALSound &snd, const func<void, pragma::audio::ALState, pragma::audio::ALState> &function);
		static void SetSource(lua::State *l, pragma::audio::ALSound &snd);

		static void SetDirectFilter(lua::State *l, pragma::audio::ALSound &snd, float gain, float gainHF, float gainLF);
		static void SetDirectFilter(lua::State *l, pragma::audio::ALSound &snd, float gain);
		static std::tuple<float, float, float> GetDirectFilter(lua::State *l, pragma::audio::ALSound &snd);
		static bool AddEffect(lua::State *l, pragma::audio::ALSound &snd, const std::string &name, float gain, float gainHF, float gainLF);
		static bool AddEffect(lua::State *l, pragma::audio::ALSound &snd, const std::string &name, float gain);
		static void SetEffectParameters(lua::State *l, pragma::audio::ALSound &snd, const std::string &name, float gain, float gainHF, float gainLF);
		static void SetEffectParameters(lua::State *l, pragma::audio::ALSound &snd, const std::string &name, float gain);
	};
};

void Lua::ALSound::register_class(luabind::class_<pragma::audio::ALSound> &classDef)
{
	classDef.def(luabind::tostring(luabind::self));
	classDef.def("Play", &pragma::audio::ALSound::Play);
	classDef.def("Stop", &pragma::audio::ALSound::Stop);
	classDef.def("SetOffset", &pragma::audio::ALSound::SetOffset);
	classDef.def("GetOffset", &pragma::audio::ALSound::GetOffset);
	classDef.def("SetPitch", &pragma::audio::ALSound::SetPitch);
	classDef.def("GetPitch", &pragma::audio::ALSound::GetPitch);
	classDef.def("SetLooping", &pragma::audio::ALSound::SetLooping);
	classDef.def("IsLooping", &pragma::audio::ALSound::IsLooping);
	classDef.def("IsPlaying", &pragma::audio::ALSound::IsPlaying);
	classDef.def("IsStopped", &pragma::audio::ALSound::IsStopped);
	classDef.def("GetState", &pragma::audio::ALSound::GetState);
	classDef.def("Rewind", &pragma::audio::ALSound::Rewind);
	classDef.def("Pause", &pragma::audio::ALSound::Pause);
	classDef.def("IsPaused", &pragma::audio::ALSound::IsPaused);
	classDef.def("SetGain", &pragma::audio::ALSound::SetGain);
	classDef.def("GetGain", &pragma::audio::ALSound::GetGain);
	classDef.def("SetPos", &pragma::audio::ALSound::SetPosition);
	classDef.def("GetPos", &pragma::audio::ALSound::GetPosition);
	classDef.def("SetVelocity", &pragma::audio::ALSound::SetVelocity);
	classDef.def("GetVelocity", &pragma::audio::ALSound::GetVelocity);
	classDef.def("SetDirection", &pragma::audio::ALSound::SetDirection);
	classDef.def("GetDirection", &pragma::audio::ALSound::GetDirection);
	classDef.def("SetRelative", &pragma::audio::ALSound::SetRelative);
	classDef.def("IsRelative", &pragma::audio::ALSound::IsRelative);
	classDef.def("GetDuration", &pragma::audio::ALSound::GetDuration);
	classDef.def("GetReferenceDistance", &pragma::audio::ALSound::GetReferenceDistance);
	classDef.def("SetReferenceDistance", &pragma::audio::ALSound::SetReferenceDistance);
	classDef.def("GetRolloffFactor", &pragma::audio::ALSound::GetRolloffFactor);
	classDef.def("SetRolloffFactor", &pragma::audio::ALSound::SetRolloffFactor);
	classDef.def("GetRoomRolloffFactor", &pragma::audio::ALSound::GetRoomRolloffFactor);
	classDef.def("SetRoomRolloffFactor", &pragma::audio::ALSound::SetRoomRolloffFactor);
	classDef.def("GetMaxDistance", &pragma::audio::ALSound::GetMaxDistance);
	classDef.def("SetMaxDistance", &pragma::audio::ALSound::SetMaxDistance);
	classDef.def("GetMinGain", &pragma::audio::ALSound::GetMinGain);
	classDef.def("SetMinGain", &pragma::audio::ALSound::SetMinGain);
	classDef.def("GetMaxGain", &pragma::audio::ALSound::GetMaxGain);
	classDef.def("SetMaxGain", &pragma::audio::ALSound::SetMaxGain);
	classDef.def("GetInnerConeAngle", &pragma::audio::ALSound::GetInnerConeAngle);
	classDef.def("SetInnerConeAngle", &pragma::audio::ALSound::SetInnerConeAngle);
	classDef.def("GetOuterConeAngle", &pragma::audio::ALSound::GetOuterConeAngle);
	classDef.def("SetOuterConeAngle", &pragma::audio::ALSound::SetOuterConeAngle);
	classDef.def("GetOuterConeGain", &pragma::audio::ALSound::GetOuterConeGain);
	classDef.def("SetOuterConeGain", &pragma::audio::ALSound::SetOuterConeGain);
	classDef.def("GetOuterConeGainHF", &pragma::audio::ALSound::GetOuterConeGainHF);
	classDef.def("SetOuterConeGainHF", &pragma::audio::ALSound::SetOuterConeGainHF);
	classDef.def("GetType", &pragma::audio::ALSound::GetType);
	classDef.def("SetType", &pragma::audio::ALSound::SetType);
	classDef.def("AddType", &pragma::audio::ALSound::AddType);
	classDef.def("CallOnStateChanged", &CallOnStateChanged);
	classDef.def("SetPriority", &pragma::audio::ALSound::SetPriority);
	classDef.def("GetPriority", &pragma::audio::ALSound::GetPriority);
	classDef.def("FadeIn", &pragma::audio::ALSound::FadeIn);
	classDef.def("FadeOut", &pragma::audio::ALSound::FadeOut);
	classDef.def("GetMaxAudibleDistance", &pragma::audio::ALSound::GetMaxAudibleDistance);
	classDef.def("GetIntensity", &pragma::audio::ALSound::GetSoundIntensity);
	classDef.def("GetSource", &pragma::audio::ALSound::GetSource);
	classDef.def("SetSource", static_cast<void (pragma::audio::ALSound::*)(pragma::ecs::BaseEntity *)>(&pragma::audio::ALSound::SetSource));
	classDef.def("SetSource", static_cast<void (*)(lua::State *, pragma::audio::ALSound &)>(&SetSource));
	classDef.def("SetRange", &pragma::audio::ALSound::SetRange);
	classDef.def("ClearRange", &pragma::audio::ALSound::ClearRange);
	classDef.def("HasRange", &pragma::audio::ALSound::HasRange);
	classDef.def("GetRange", &pragma::audio::ALSound::GetRange);
	classDef.def("GetRangeOffsets", &pragma::audio::ALSound::GetRangeOffsets);
	classDef.def("SetFadeInDuration", &pragma::audio::ALSound::SetFadeInDuration);
	classDef.def("SetFadeOutDuration", &pragma::audio::ALSound::SetFadeOutDuration);
	classDef.def("GetFadeInDuration", &pragma::audio::ALSound::GetFadeInDuration);
	classDef.def("GetFadeOutDuration", &pragma::audio::ALSound::GetFadeOutDuration);

	classDef.def("SetOrientation", &pragma::audio::ALSound::SetOrientation);
	classDef.def("GetOrientation", &pragma::audio::ALSound::GetOrientation);
	classDef.def("SetDopplerFactor", &pragma::audio::ALSound::SetDopplerFactor);
	classDef.def("GetDopplerFactor", &pragma::audio::ALSound::GetDopplerFactor);
	classDef.def("SetLeftStereoAngle", &pragma::audio::ALSound::SetLeftStereoAngle);
	classDef.def("GetLeftStereoAngle", &pragma::audio::ALSound::GetLeftStereoAngle);
	classDef.def("SetRightStereoAngle", &pragma::audio::ALSound::SetRightStereoAngle);
	classDef.def("GetRightStereoAngle", &pragma::audio::ALSound::GetRightStereoAngle);
	classDef.def("SetAirAbsorptionFactor", &pragma::audio::ALSound::SetAirAbsorptionFactor);
	classDef.def("GetAirAbsorptionFactor", &pragma::audio::ALSound::GetAirAbsorptionFactor);
	classDef.def("SetGainAuto", &pragma::audio::ALSound::SetGainAuto);
	classDef.def("GetGainAuto", &pragma::audio::ALSound::GetGainAuto);
	classDef.def("SetDirectFilter", static_cast<void (*)(lua::State *, pragma::audio::ALSound &, float, float, float)>(&SetDirectFilter));
	classDef.def("SetDirectFilter", static_cast<void (*)(lua::State *, pragma::audio::ALSound &, float)>(&SetDirectFilter));
	classDef.def("GetDirectFilter", &GetDirectFilter);
	classDef.def("AddEffect", static_cast<bool (*)(lua::State *, pragma::audio::ALSound &, const std::string &, float, float, float)>(&AddEffect));
	classDef.def("AddEffect", static_cast<bool (*)(lua::State *, pragma::audio::ALSound &, const std::string &, float)>(&AddEffect));
	classDef.def("RemoveEffect", &pragma::audio::ALSound::RemoveEffect);
	classDef.def("SetEffectParameters", static_cast<bool (*)(lua::State *, pragma::audio::ALSound &, const std::string &, float, float, float)>(&AddEffect));
	classDef.def("SetEffectParameters", static_cast<bool (*)(lua::State *, pragma::audio::ALSound &, const std::string &, float)>(&AddEffect));

	classDef.def("SetGainRange", &pragma::audio::ALSound::SetGainRange);
	classDef.def("GetGainRange", &pragma::audio::ALSound::GetGainRange);
	classDef.def("SetDistanceRange", &pragma::audio::ALSound::SetDistanceRange);
	classDef.def("GetDistanceRange", &pragma::audio::ALSound::GetDistanceRange);
	classDef.def("SetConeAngles", &pragma::audio::ALSound::SetConeAngles);
	classDef.def("GetConeAngles", &pragma::audio::ALSound::GetConeAngles);
	classDef.def("GetOuterConeGains", &pragma::audio::ALSound::GetOuterConeGains);
	classDef.def("SetStereoAngles", &pragma::audio::ALSound::SetStereoAngles);
	classDef.def("GetStereoAngles", &pragma::audio::ALSound::GetStereoAngles);
	classDef.def("SetOuterConeGains", &pragma::audio::ALSound::SetOuterConeGains);
	classDef.def("GetDirectGainHFAuto", &pragma::audio::ALSound::GetDirectGainHFAuto);
	classDef.def("GetSendGainAuto", &pragma::audio::ALSound::GetSendGainAuto);
	classDef.def("GetSendGainHFAuto", &pragma::audio::ALSound::GetSendGainHFAuto);
	classDef.def("SetEffectGain", &pragma::audio::ALSound::SetEffectGain);
	classDef.def("GetRolloffFactors", &pragma::audio::ALSound::GetRolloffFactors);
	classDef.def("SetRolloffFactors", &pragma::audio::ALSound::SetRolloffFactors);
}

CallbackHandle Lua::ALSound::CallOnStateChanged(lua::State *l, pragma::audio::ALSound &snd, const func<void, pragma::audio::ALState, pragma::audio::ALState> &function) { return snd.AddLuaCallback("OnStateChanged", function); }

void Lua::ALSound::SetSource(lua::State *, pragma::audio::ALSound &snd) { snd.SetSource(nullptr); }
void Lua::ALSound::SetDirectFilter(lua::State *l, pragma::audio::ALSound &snd, float gain, float gainHF, float gainLF) { snd.SetDirectFilter({gain, gainHF, gainLF}); }
void Lua::ALSound::SetDirectFilter(lua::State *l, pragma::audio::ALSound &snd, float gain) { snd.SetDirectFilter({gain}); }
std::tuple<float, float, float> Lua::ALSound::GetDirectFilter(lua::State *l, pragma::audio::ALSound &snd)
{
	auto &filter = snd.GetDirectFilter();
	return {filter.gain, filter.gainHF, filter.gainLF};
}
bool Lua::ALSound::AddEffect(lua::State *l, pragma::audio::ALSound &snd, const std::string &name, float gain, float gainHF, float gainLF) { return snd.AddEffect(name, {gain, gainHF, gainLF}); }
bool Lua::ALSound::AddEffect(lua::State *l, pragma::audio::ALSound &snd, const std::string &name, float gain) { return snd.AddEffect(name, {gain}); }
void Lua::ALSound::SetEffectParameters(lua::State *l, pragma::audio::ALSound &snd, const std::string &name, float gain, float gainHF, float gainLF) { snd.SetEffectParameters(name, {gain, gainHF, gainLF}); }
void Lua::ALSound::SetEffectParameters(lua::State *l, pragma::audio::ALSound &snd, const std::string &name, float gain) { snd.SetEffectParameters(name, {gain}); }
